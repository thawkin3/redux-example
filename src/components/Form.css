.search-form {
  margin: 10px auto 30px auto;
  position: relative;
}

.search-label {
  display: block;
  font-size: 75%;
}
