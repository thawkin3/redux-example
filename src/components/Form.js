import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions/index';
import './Form.css';

export class Form extends Component {
    constructor(props) {
        super(props);

        this.state = { username: '' };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange(e) {
        this.setState({ username: e.target.value });
    }

    handleSubmit(e) {
        e.preventDefault();
        const { username } = this.state;
        if (username) {
            this.props.fetchGitHubUser(username);
            this.setState({ username: '' });
        }
    };

    render() {
        const { username } = this.state;
        return (
            <form onSubmit={this.handleSubmit}>
                <input
                    type="text"
                    value={username}
                    onChange={this.handleChange}
                    placeholder="Github username"
                    required
                />
                <button type="submit">Add card</button>
            </form>
        );
    };
}

export default connect(null, actions)(Form);
