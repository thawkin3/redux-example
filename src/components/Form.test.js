import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Form } from './Form';

let wrapper, props;

describe('Form', () => {
    beforeEach(() => {
        props = {
            ajaxRequestInProgress: false,
            fetchGitHubUser: () => {},
        };
        wrapper = shallow(<Form {...props} />);
    });

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Form {...props} />, div);
    });

    it('matches the snapshot', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('renders only one Form', () => {
        expect(wrapper.length).toEqual(1);
    });

    it('renders a form with the class search-form', () => {
        expect(wrapper.find('form.search-form').length).toEqual(1);
    })

    it('renders a button, label, and input element', () => {
        expect(wrapper.find('button').length).toEqual(1);
        expect(wrapper.find('label').length).toEqual(1);
        expect(wrapper.find('input').length).toEqual(1);
    });

    it('renders a p element with the class searching when ajaxRequestInProgress is true', () => {
        expect(wrapper.find('p.searching').length).toEqual(0);
        wrapper.setProps({ ajaxRequestInProgress: true });
        expect(wrapper.find('p.searching').length).toEqual(1);
    });

    it('responds to the submit event on the form when there is a username provided', () => {
        const spy = jest.spyOn(Form.prototype, 'handleSubmit');
        const spy2 = jest.spyOn(props, 'fetchGitHubUser');
        wrapper = shallow(<Form {...props} />);
        wrapper.setState({ username: 'thawkin3' });
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy2).toHaveBeenCalledTimes(1);
        expect(wrapper.state().username).toEqual('');
        spy.mockRestore();
        spy2.mockRestore();
    });

    it('responds to the submit event on the form when there is not a username provided', () => {
        const spy = jest.spyOn(Form.prototype, 'handleSubmit');
        const spy2 = jest.spyOn(props, 'fetchGitHubUser');
        wrapper = shallow(<Form {...props} />);
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy2).toHaveBeenCalledTimes(0);
        spy.mockRestore();
        spy2.mockRestore();
    });

    it('responds to the change event on the input', () => {
        const spy = jest.spyOn(Form.prototype, 'handleChange');
        wrapper = shallow(<Form {...props} />);
        wrapper.find('input').simulate('change', { target: { value: 'thawkin3' } });
        expect(spy).toHaveBeenCalledTimes(1);
        spy.mockRestore();
    });
});
