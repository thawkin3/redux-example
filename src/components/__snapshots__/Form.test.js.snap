// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches the snapshot 1`] = `
<form
  className="search-form"
  onSubmit={[Function]}
>
  <label
    className="search-label"
    htmlFor="search"
  >
    Search for a GitHub user
  </label>
  <input
    disabled={false}
    id="search"
    name="search"
    onChange={[Function]}
    placeholder="Github username"
    required={true}
    type="text"
    value=""
  />
  <button
    type="submit"
  >
    Add card
  </button>
</form>
`;
